lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,477 @@
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchReadFirstLane.cpp
+ * @brief LLPC source file: contains declaration and implementation of class lgc::PatchReadFirstLane.
+ ***********************************************************************************************************************
+ */
+#include "lgc/Builder.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PipelineState.h"
+#include "llvm/Analysis/LegacyDivergenceAnalysis.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/Debug.h"
+#include <deque>
+
+#define DEBUG_TYPE "lgc-patch-read-first-lane"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace {
+class PatchReadFirstLane final : public FunctionPass {
+public:
+  PatchReadFirstLane();
+
+  virtual bool runOnFunction(Function &function) override;
+  void getAnalysisUsage(AnalysisUsage &analysisUsage) const override;
+
+  static char ID; // ID of this pass
+
+private:
+  PatchReadFirstLane(const PatchReadFirstLane &) = delete;
+  PatchReadFirstLane &operator=(const PatchReadFirstLane &) = delete;
+
+  bool liftReadFirstLane(Function &function);
+  void collectAssumeUniforms(BasicBlock *block, const SmallVectorImpl<Instruction *> &initialReadFirstLanes);
+  void findBestInsertLocation(const SmallVectorImpl<Instruction *> &initialReadFirstLanes);
+  bool isAllUsesDependCandidate(Instruction *candidateInsertLocatin,
+                                const DenseSet<Instruction *> &nonEarliestDivergentUses,
+                                DenseSet<Instruction *> &visitedInsts);
+  bool isAllUsersAssumedUniform(Instruction *inst);
+  void applyReadFirstLane(Instruction *inst, BuilderBase &builder);
+  void addLocation(Instruction *inst);
+
+  // We only support to apply amdgcn_readfirstlane on float or int type
+  // TODO: Support various types when backend work is ready
+  bool isSupportedType(Instruction *inst) { return inst->getType()->isFloatTy() || inst->getType()->isIntegerTy(32); }
+
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis
+  TargetTransformInfo *m_targetTransformInfo;     // The target tranform info to determine stop propagation
+  DenseMap<Instruction *, SmallVector<Instruction *, 2>>
+      m_canAssumeUniformDivergentUsesMap; // The map key is an instruction `I` that can be assumed uniform. The map
+                                          // value is a vector of instructions that we can apply readfirstlane on them.
+                                          // An empty vector means the instruction `I` is an insert location.
```

Comment:
Suggestion for improving the comment slightly (I don't know how to check the line length on GitHub):
```suggestion
      m_canAssumeUniformDivergentUsesMap; // The map key is an instruction `I` that can be assumed uniform.
                                          // That is, we can apply readfirstlane to the result of `I` and remain correct. If the map
                                          // value vector is non-empty, it contains a list of instructions that we can apply
                                          // readfirstlane on to achieve the same effect as a readfirstlane on `I`.
                                          // An empty vector means that it is not possible to life a readfirstlane beyond `I`.
```